Add prompt caching

Yukinobu-Mine opened the pull request:
*Issue #, if available:*
Closes #805 

*Description of changes:*
- Reflect the cost of cache read/write to the price calculation. https://github.com/aws-samples/bedrock-chat/pull/891/commits/caa353fcea0689b9f7d6728871583b0a1000f1b7
- Implement prompt caching. https://github.com/aws-samples/bedrock-chat/pull/891/commits/501d6b59b2af664c0c14196c4f1d559c59de0def https://github.com/aws-samples/bedrock-chat/pull/891/commits/3be9ba82bc8a95af9fa0768299d51fc61f255015

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Yukinobu-Mine committed:
Reflect the cost of cache read/write to the price calculation.

Yukinobu-Mine committed:
Implement prompt caching.
- Backend
  - Add a field `use_prompt_caching` to models and schemas of custom bot.
    - `BotModel`
    - `BotInput`
    - `BotOutput`
    - `BotModifyInput`
    - `BotModifyOutput`
  - Add a column `UsePromptCaching` to DynamoDB table `BotTableV3`.
  - Use prompt caching if enabled and the model supports it.
- Frontend
  - Add a field `usePromptCaching` to schemas of custom bot.
    - `BotDetails`
    - `RegisterBotRequest`
    - `RegisterBotResponse`
    - `UpdateBotRequest`
    - `UpdateBotResponse`
  - Add 'Prompt Caching' section to `BotKbEditPage`.

Yukinobu-Mine committed:
[Debug]Print token count and price when received `STREAMING_END`.

Yukinobu-Mine committed:
Reformat modified python codes.
- `backend/app/config.py`
- `backend/app/bedrock.py`
- `backend/app/repositories/custom_bot.py`
- `backend/app/usecases/chat.py`

Yukinobu-Mine committed:
Use `ExpandableDrawerGroup` for prompt caching settings.

Yukinobu-Mine committed:
Revert "[Debug]Print token count and price when received `STREAMING_END`."

This reverts commit ba5e5845971c85097d005f70cf2682a71ef1ae4a.

Yukinobu-Mine committed:
Refactor data structure of agent settings.

statefb commented on frontend/src/features/knowledgeBase/pages/BotKbEditPage.tsx:
```diff
@@ -107,6 +107,7 @@ const BotKbEditPage: React.FC = () => {
     defaultGenerationConfig.reasoningParams?.budgetTokens ??
       EDGE_GENERATION_PARAMS.budgetTokens.MIN
   );
+  const [usePromptCaching, setUsePromptCaching] = useState<boolean>();
```
`useXXX`というネーミングがhooksと誤解しやすいので、`promptCachingEnabled`など検討ください！

Yukinobu-Mine replied:
修正しました。

statefb commented on frontend/src/features/knowledgeBase/pages/BotKbEditPage.tsx:
```diff
@@ -1205,40 +1207,7 @@ const BotKbEditPage: React.FC = () => {
-
-          return baseTool;
-        }),
+        tools,
```
refactorありがとうございます！

statefb commented on backend/app/repositories/models/custom_bot.py:
```diff
@@ -388,6 +387,7 @@ class BotModel(BaseModel):
     generation_params: GenerationParamsModel
     agent: AgentModel
     knowledge: KnowledgeModel
+    use_prompt_caching: bool | None = None
```
(ModelだけでなくSchemaにも共通して言えることですが、)なるべく状態を減らすため (bool | Noneではなく) boolのみにすることをご検討ください！（既存のbotはすべてTrueにfall back）

Yukinobu-Mine replied:
以下について修正しました。
- `BotModel`
- `BotInput`
- `BotOutput`

`BotModifyInput` と `BotModifyOutput` については、`promptCachingEnabled` を省略した場合は「更新しない」という動作にするため、bool | None のままにしています。

statefb commented on backend/app/repositories/models/custom_bot.py:
```diff
@@ -468,6 +468,9 @@ def is_agent_enabled(self) -> bool:
         return len(self.agent.tools) > 0 or self.has_knowledge()
 
+    def is_prompt_caching_enabled(self) -> bool:
+        return self.use_prompt_caching != False
```
bool | None -> boolにすれば不要？

Yukinobu-Mine replied:
削除しました。

Yukinobu-Mine committed:
Rename `usePromptCaching` to `promptCachingEnabled`.
- `use_prompt_caching` -> `prompt_caching_enabled`
- Change `BotModel.prompt_caching_enabled` to a non-nullable type.

Yukinobu-Mine commented:
@statefb コメントありがとうございます。修正したのでご確認下さい。

Yukinobu-Mine committed:
Change `BotModifyInput.prompt_caching_enabled` to a non-nullable type.

Yukinobu-Mine committed:
Add `token_count` and `price` to the payload of `STREAMING_END` notification.

statefb reviewed (approved):
LGTM!
